File: snrna/integration/plot_class_integration_summary.py
"""Class-level summary of cross-species integration strength, finch x chicken vs finch x mouse.

Tests the write-up claim that all finch cell classes integrate well with chicken, while with
mouse non-neuronal cells integrate strongly, GABAergic neurons moderately and glutamatergic
neurons weakly. Reads the two full-suite hybrid-label composites (GSI + SAMap + CCA + SATURN):
  composite_scoring/results/gg_adult_hybrid/composite_calls.csv   (assemble_gg_adult_hybrid.py)
  composite_scoring/results/yao_adult_hybrid/composite_calls.csv  (assemble_yao_adult_hybrid.py)

Two per-cluster readouts, each summarised by finch cell class (Non-neuronal / GABAergic /
Glutamatergic, from the celltype_hybrid name prefix):
  top row     composite confidence -- the Zaremba magnitude channel evaluated at the rank-
              aggregated winner (composite_score.summarise); "is the best match strong?"
  bottom row  method agreement -- how many of the 4 methods independently pick the same
              reference label as their own top call; "do the methods agree on WHAT it is?"
Points are individual finch clusters; bars are class medians. Within-pair class differences
are tested by one-sided Mann-Whitney U (direction: Non-neuronal > GABA > Glut) and the
chicken-vs-mouse drop per class by paired Wilcoxon over shared finch clusters.
"""
from pathlib import Path
import matplotlib
matplotlib.use("Agg")
import matplotlib.pyplot as plt
from matplotlib.lines import Line2D
import numpy as np
import pandas as pd
from scipy.stats import mannwhitneyu, wilcoxon

plt.rcParams.update({"font.family": "Arial", "pdf.fonttype": 42, "ps.fonttype": 42,
                     "font.size": 7, "axes.linewidth": 0.6})

BASE = Path("/private/groups/colquittlab/song-system-grn/snrna/integration/composite_scoring/results")
OUT = BASE / "class_summary"
OUT.mkdir(exist_ok=True)

PAIRS = [("gg_adult_hybrid", "Finch × chicken"), ("yao_adult_hybrid", "Finch × mouse")]
CLASSES = ["Non-neuronal", "GABAergic", "Glutamatergic"]
COLORS = {"Non-neuronal": "#1b9e77", "GABAergic": "#7570b3", "Glutamatergic": "#d95f02"}
INK, MUTED = "#222222", "#777777"


def cell_class(q):
    if q.startswith("GABA"):
        return "GABAergic"
    if q.startswith("Glut"):
        return "Glutamatergic"
    return "Non-neuronal"


def load(tag):
    D = pd.read_csv(BASE / tag / "composite_calls.csv", index_col=0)
    D["cls"] = [cell_class(q) for q in D.index]
    D["agree_frac"] = D.methods_agreeing_on_own_top / D.n_methods
    return D


def pstar(p):
    return "***" if p < 0.001 else "**" if p < 0.01 else "*" if p < 0.05 else "ns"


data = {tag: load(tag) for tag, _ in PAIRS}

# ---- per-class summary table -----------------------------------------------------------
rows = []
for tag, label in PAIRS:
    D = data[tag]
    for c in CLASSES:
        d = D[D.cls == c]
        rows.append(dict(pair=label, cell_class=c, n_clusters=len(d),
                         confidence_median=d.confidence.median(), confidence_mean=d.confidence.mean(),
                         frac_high_tier=(d.confidence_tier == "high").mean(),
                         frac_low_tier=(d.confidence_tier == "low").mean(),
                         agree_frac_mean=d.agree_frac.mean(),
                         frac_all_methods_agree=(d.methods_agreeing_on_own_top == d.n_methods).mean(),
                         frac_le1_method_agree=(d.methods_agreeing_on_own_top <= 1).mean()))
summ = pd.DataFrame(rows)
summ.to_csv(OUT / "class_summary_table.csv", index=False)

tests = []
for tag, label in PAIRS:
    D = data[tag]
    for met in ["confidence", "agree_frac"]:
        g = {c: D.loc[D.cls == c, met].values for c in CLASSES}
        for a, b in [("Non-neuronal", "GABAergic"), ("GABAergic", "Glutamatergic"), ("Non-neuronal", "Glutamatergic")]:
            p = mannwhitneyu(g[a], g[b], alternative="greater").pvalue
            tests.append(dict(pair=label, metric=met, test=f"{a} > {b}", p=p))
G, Y = data["gg_adult_hybrid"], data["yao_adult_hybrid"]
for c in CLASSES:
    idx = [q for q in G.index if G.cls[q] == c and q in Y.index]
    for met in ["confidence", "agree_frac"]:
        try:
            p = wilcoxon(G.loc[idx, met], Y.loc[idx, met], alternative="greater").pvalue
        except ValueError:
            p = np.nan
        tests.append(dict(pair="chicken vs mouse (paired)", metric=met, test=f"{c}: chicken > mouse", p=p))
tests = pd.DataFrame(tests)
tests.to_csv(OUT / "class_summary_tests.csv", index=False)
print(summ.round(3).to_string(index=False))
print()
print(tests.assign(p=tests.p.map(lambda x: f"{x:.2g}")).to_string(index=False))

# ---- figure -----------------------------------------------------------------------------
METRICS = [("confidence", "Composite confidence\n(match strength at winner)", (0, 1.0)),
           ("agree_frac", "Method agreement\n(fraction of 4 methods sharing top call)", (0, 1.0))]
fig, axes = plt.subplots(2, 2, figsize=(4.6, 4.2), sharey="row")
rng = np.random.default_rng(0)
for j, (tag, label) in enumerate(PAIRS):
    D = data[tag]
    for i, (met, ylabel, ylim) in enumerate(METRICS):
        ax = axes[i, j]
        for k, c in enumerate(CLASSES):
            v = D.loc[D.cls == c, met].values
            med = np.median(v) if met == "confidence" else np.mean(v)
            ax.bar(k, med, width=0.62, color=COLORS[c], alpha=0.28, linewidth=0, zorder=1)
            ax.hlines(med, k - 0.31, k + 0.31, color=COLORS[c], linewidth=1.6, zorder=3)
            if met == "agree_frac":   # discrete 0.25 steps: fan out horizontally, wrap rows of 6
                for val in np.unique(v):
                    n = (v == val).sum()
                    for r0 in range(0, n, 6):
                        m = min(6, n - r0)
                        xs = k + (np.arange(m) - (m - 1) / 2) * 0.075
                        ys = np.full(m, val) + 0.045 * (r0 // 6)
                        ax.scatter(xs, ys, s=9, color=COLORS[c], edgecolor="white",
                                   linewidth=0.4, zorder=4)
            else:
                xs = k + rng.uniform(-0.18, 0.18, len(v))
                ax.scatter(xs, v, s=9, color=COLORS[c], edgecolor="white", linewidth=0.4, zorder=4)
        # pairwise brackets: adjacent classes
        g = {c: D.loc[D.cls == c, met].values for c in CLASSES}
        for k, (a, b) in enumerate([("Non-neuronal", "GABAergic"), ("GABAergic", "Glutamatergic")]):
            p = mannwhitneyu(g[a], g[b], alternative="greater").pvalue
            y = 1.08
            ax.plot([k + 0.05, k + 0.05, k + 0.95, k + 0.95], [y - 0.02, y, y, y - 0.02],
                    color=INK, linewidth=0.6, clip_on=False)
            ax.text(k + 0.5, y + 0.01, pstar(p), ha="center", va="bottom", fontsize=6.5, color=INK)
        ax.set_ylim(0, 1.0)
        ax.set_xlim(-0.55, 2.55)
        ax.set_xticks(range(3))
        ns_ = [int((D.cls == c).sum()) for c in CLASSES]
        ax.set_xticklabels([f"Non-\nneuronal\n(n={ns_[0]})", f"GABA\n(n={ns_[1]})", f"Glut\n(n={ns_[2]})"],
                           fontsize=6.5)
        ax.set_yticks([0, 0.25, 0.5, 0.75, 1.0])
        ax.spines[["top", "right"]].set_visible(False)
        ax.tick_params(length=2, width=0.6, colors=INK, labelsize=6.5)
        ax.grid(axis="y", color="#e6e6e6", linewidth=0.5, zorder=0)
        ax.set_axisbelow(True)
        if met == "confidence":
            ax.axhline(0.5, color=MUTED, linewidth=0.6, linestyle=(0, (3, 2)), zorder=2)
            if j == 1:
                ax.text(2.55, 0.5, " high-\n tier", va="center", ha="left", fontsize=5.5, color=MUTED)
        if j == 0:
            ax.set_ylabel(ylabel, fontsize=7)
        if i == 0:
            ax.set_title(label, fontsize=8, pad=16, color=INK)

fig.text(0.5, 0.01,
         "Points: finch clusters. Bars: class median (top) or mean (bottom).\n"
         "Brackets: one-sided Mann-Whitney U, left > right; * p<0.05, ** p<0.01, *** p<0.001, ns p>0.05.",
         ha="center", va="bottom", fontsize=5.5, color=MUTED)
fig.subplots_adjust(left=0.16, right=0.95, top=0.89, bottom=0.15, hspace=0.6, wspace=0.12)
for ext in ("pdf", "png"):
    fig.savefig(OUT / f"class_integration_summary.{ext}", dpi=300)
print(f"\nwrote {OUT}")

File: snrna/integration/plot_class_integration_summary.R
## ---------------------------------------------------------------------------
## Class-level summary of cross-species integration strength,
## finch x chicken vs finch x mouse.
##
## Tests the write-up claim that all finch cell classes integrate well with
## chicken, while with mouse non-neuronal cells integrate strongly, GABAergic
## neurons moderately and glutamatergic neurons weakly. Reads the two full-suite
## hybrid-label composites (GSI + SAMap + CCA + SATURN):
##   composite_scoring/results/gg_adult_hybrid/composite_calls.csv
##   composite_scoring/results/yao_adult_hybrid/composite_calls.csv
##
## Two per-cluster readouts, summarised by finch cell class (Non-neuronal /
## GABAergic / Glutamatergic, from the celltype_hybrid name prefix):
##   top row     composite confidence -- the Zaremba magnitude channel at the
##               rank-aggregated winner; "is the best match strong?"
##   bottom row  method agreement -- how many of the 4 methods independently pick
##               the same reference label as their own top call.
## Points are finch clusters; bars are the class median (top) or mean (bottom).
## Within-pair class differences: one-sided Mann-Whitney U (Non-neuronal > GABA
## > Glut), normal approximation with continuity correction. Chicken-vs-mouse
## drop per class: paired one-sided Wilcoxon over shared finch clusters.
##
## Writes class_summary_table.csv / class_summary_tests.csv (tracked) and the
## figure (PDF + PNG, gitignored) to composite_scoring/results/class_summary/.
## ---------------------------------------------------------------------------

suppressMessages({
  library(here)
  library(dplyr)
  library(tidyr)
  library(ggplot2)
  library(patchwork)
})
source(here::here("config/figure_theme.R"))

RES <- here::here("snrna/integration/composite_scoring/results")
OUT <- file.path(RES, "class_summary"); dir.create(OUT, showWarnings = FALSE)

PAIRS <- c(gg_adult_hybrid = "Finch × chicken", yao_adult_hybrid = "Finch × mouse")
CLASSES <- c("Non-neuronal", "GABAergic", "Glutamatergic")
CLASS_COL <- setNames(FIG_PAL[1:3], CLASSES)   # fixed slots, all-pairs safe
METRICS <- c(confidence = "Composite confidence\n(match strength at winner)",
             agree_frac = "Method agreement\n(fraction of 4 methods)")

cell_class <- function(q) ifelse(grepl("^GABA", q), "GABAergic",
                          ifelse(grepl("^Glut", q), "Glutamatergic", "Non-neuronal"))

D <- bind_rows(lapply(names(PAIRS), function(tag) {
  read.csv(file.path(RES, tag, "composite_calls.csv"), check.names = FALSE) %>%
    transmute(pair = PAIRS[[tag]], tag, query, confidence, confidence_tier,
              methods_agreeing_on_own_top, n_methods,
              agree_frac = methods_agreeing_on_own_top / n_methods,
              cls = factor(cell_class(query), CLASSES))
}))

## --- Tables -------------------------------------------------------------------
summ <- D %>% group_by(pair, cell_class = cls) %>%
  summarise(n_clusters = n(),
            confidence_median = median(confidence), confidence_mean = mean(confidence),
            frac_high_tier = mean(confidence_tier == "high"),
            frac_low_tier = mean(confidence_tier == "low"),
            agree_frac_mean = mean(agree_frac),
            frac_all_methods_agree = mean(methods_agreeing_on_own_top == n_methods),
            frac_le1_method_agree = mean(methods_agreeing_on_own_top <= 1), .groups = "drop")
write.csv(summ, file.path(OUT, "class_summary_table.csv"), row.names = FALSE)

mwu <- function(a, b) wilcox.test(a, b, alternative = "greater", exact = FALSE, correct = TRUE)$p.value
COMPS <- list(c("Non-neuronal", "GABAergic"), c("GABAergic", "Glutamatergic"),
              c("Non-neuronal", "Glutamatergic"))
tests <- bind_rows(
  lapply(names(PAIRS), function(tag) bind_rows(lapply(names(METRICS), function(met) bind_rows(lapply(COMPS, function(cp) {
    d <- D[D$tag == tag, ]
    tibble(pair = PAIRS[[tag]], metric = met, test = sprintf("%s > %s", cp[1], cp[2]),
           p = mwu(d[[met]][d$cls == cp[1]], d[[met]][d$cls == cp[2]]))
  }))))),
  lapply(CLASSES, function(c) bind_rows(lapply(names(METRICS), function(met) {
    G <- D[D$tag == "gg_adult_hybrid" & D$cls == c, ]; Y <- D[D$tag == "yao_adult_hybrid", ]
    idx <- G$query[G$query %in% Y$query]
    p <- tryCatch(wilcox.test(G[[met]][match(idx, G$query)], Y[[met]][match(idx, Y$query)],
                              paired = TRUE, alternative = "greater")$p.value,
                  error = function(e) NA_real_)
    tibble(pair = "chicken vs mouse (paired)", metric = met, test = sprintf("%s: chicken > mouse", c), p = p)
  })))
)
write.csv(tests, file.path(OUT, "class_summary_tests.csv"), row.names = FALSE)
print(as.data.frame(summ %>% mutate(across(where(is.numeric), ~ round(.x, 3)))), row.names = FALSE)
print(as.data.frame(tests %>% mutate(p = signif(p, 3))), row.names = FALSE)

## --- Figure ---------------------------------------------------------------------
## Full p-value rather than a significance star, per direct instruction --
## scientific notation below 0.001 (where a fixed-decimal value would round to
## 0.000), three decimal places otherwise.
pfmt <- function(p) ifelse(p < 0.001, sprintf("p=%.1e", p), sprintf("p=%.3f", p))
long <- D %>% pivot_longer(c(confidence, agree_frac), names_to = "metric", values_to = "value") %>%
  mutate(metric = factor(metric, names(METRICS)), pair = factor(pair, PAIRS), x = as.integer(cls))

## point positions: jitter for the continuous metric, deterministic fan-out in rows
## of six for the discrete agreement fractions (0.25 steps)
set.seed(0)
pts <- long %>% group_by(pair, metric, cls, value) %>%
  mutate(k = row_number() - 1L, n = n(),
         row_i = k %/% 6L, m = pmin(6L, n - 6L * row_i), j = k %% 6L,
         xp = ifelse(metric == "agree_frac", x + (j - (m - 1) / 2) * 0.075, x + runif(n(), -0.18, 0.18)),
         yp = ifelse(metric == "agree_frac", value + 0.045 * row_i, value)) %>% ungroup()
bars <- long %>% group_by(pair, metric, cls, x) %>%
  summarise(stat = if (first(metric) == "confidence") median(value) else mean(value), .groups = "drop")
ns <- D %>% count(pair, cls) %>% mutate(pair = factor(pair, PAIRS))
xlab_df <- ns %>% mutate(lab = sprintf("%s\n(n=%d)", c("Non-\nneuronal", "GABA", "Glut")[as.integer(cls)], n))
brk <- tests %>% filter(pair %in% PAIRS, test %in% c("Non-neuronal > GABAergic", "GABAergic > Glutamatergic")) %>%
  mutate(pair = factor(pair, PAIRS), metric = factor(metric, names(METRICS)),
         k = ifelse(test == "Non-neuronal > GABAergic", 1, 2), lab = pfmt(p), y = 1.11)

## facet_grid only ever draws the y-axis on its left column and the x-axis on
## its bottom row, even with scales = "free" (verified: ggplot2 does not repeat
## axes to interior/opposite panels in a grid facet -- that needs ggh4x or
## lemon, neither installed here). Built as four independent, unfaceted
## ggplots instead and assembled with patchwork, so every one draws its own
## full x/y axis, per direct instruction.
panel_plot <- function(met, pr, show_title, show_ylab) {
  b <- bars %>% filter(metric == met, pair == pr)
  pt <- pts %>% filter(metric == met, pair == pr)
  bk <- brk %>% filter(metric == met, pair == pr)
  g <- ggplot()
  if (met == "confidence")
    g <- g + geom_hline(aes(yintercept = 0.5), colour = FIG_INK_MUTED, linewidth = 0.3, linetype = "22")
  g <- g +
    geom_col(data = b, aes(x, stat, fill = cls), width = 0.62, alpha = 0.28) +
    geom_segment(data = b, aes(x = x - 0.31, xend = x + 0.31, y = stat, yend = stat, colour = cls), linewidth = 1.1) +
    geom_point(data = pt, aes(xp, yp, fill = cls), shape = 21, size = 1.1, colour = "white", stroke = 0.25) +
    geom_segment(data = bk, aes(x = k + 0.05, xend = k + 0.05, y = y - 0.02, yend = y), linewidth = 0.3, colour = FIG_INK_PRIMARY) +
    geom_segment(data = bk, aes(x = k + 0.95, xend = k + 0.95, y = y - 0.02, yend = y), linewidth = 0.3, colour = FIG_INK_PRIMARY) +
    geom_segment(data = bk, aes(x = k + 0.05, xend = k + 0.95, y = y, yend = y), linewidth = 0.3, colour = FIG_INK_PRIMARY) +
    ## Rotated: at this panel width two side-by-side horizontal p-value strings
    ## no longer fit within their own bracket's span and visually merge --
    ## vertical text only needs the bracket-span WIDTH for its stroke width,
    ## not for its string length, so it fits regardless.
    geom_text(data = bk, aes(k + 0.5, y + 0.03, label = lab), size = fig_pt(5.5), colour = FIG_INK_PRIMARY,
             angle = 90, hjust = 0, vjust = 0.5) +
    scale_fill_manual(values = CLASS_COL, guide = "none") +
    scale_colour_manual(values = CLASS_COL, guide = "none") +
    scale_x_continuous(breaks = 1:3, labels = c("Non-\nneuronal", "GABA", "Glut"), expand = expansion(add = 0.55)) +
    scale_y_continuous(breaks = seq(0, 1, 0.25), limits = c(0, 2.0), expand = expansion(0)) +
    coord_cartesian(clip = "off") +
    labs(x = NULL, y = if (show_ylab) METRICS[[met]] else NULL,
        title = if (show_title) pr else NULL) +
    theme_fig(base_size = 6.5) +
    theme(panel.grid.major.x = element_blank(), axis.line.y = element_line(colour = FIG_AXIS, linewidth = 0.3),
          plot.title = element_text(size = FIG_PT_AXIS_TITLE, hjust = 0.5))
  g
}

p <- (panel_plot("confidence", PAIRS[["gg_adult_hybrid"]], TRUE, TRUE) +
      panel_plot("confidence", PAIRS[["yao_adult_hybrid"]], TRUE, FALSE) +
      panel_plot("agree_frac", PAIRS[["gg_adult_hybrid"]], FALSE, TRUE) +
      panel_plot("agree_frac", PAIRS[["yao_adult_hybrid"]], FALSE, FALSE)) +
  plot_layout(ncol = 2) +
  plot_annotation(
    ## Short lines throughout -- at this panel width a sentence-per-line
    ## caption overran the right edge.
    caption = paste0("Points: finch clusters\n(non-neuronal n=", ns$n[1], ", GABA n=", ns$n[2], ", Glut n=", ns$n[3], ").\n",
                     "Bars: class median (top)\nor mean (bottom).\n",
                     "Dashed: high-confidence\ntier (0.5).\n",
                     "Brackets: one-sided\nMann-Whitney U,\nleft > right, p shown."),
    theme = theme(plot.caption = element_text(colour = FIG_INK_MUTED, size = 6.5 - 2.4, hjust = 0,
                                              family = FIG_FONT)))

fig_save(p, file.path(OUT, "class_integration_summary"), width = 2.66, height = 4.9)
message("wrote ", OUT)
